Write transformed kernels to file

arporter opened the pull request:
This PR is for #201 and adds functionality to write transformed kernels to file, as well as ensuring they are appropriately named with the PSy layer updated to match.

rupertford commented on src/psyclone/configuration.py:
```diff
@@ -459,14 +473,58 @@ def filename(self):
+        to file.
+
+        :param str value: one of VALID_KERNEL_NAMING_SCHEMES.
+        :raises ValueError: if the supplied value is not a recognised \
```
Firstly, I think it is right that this is a value error rather than a configurationerror but I thought I'd ask anyway seeing as most errors in this file are the latter.

Secondly, this error is not trapped in generate.py where the value is set. The posibility of an incorrect input should be handled there.

arporter replied:
Oops, well spotted. I've added a catch for the `ValueError` (which I chose to raise since it's an error in setting a value in the Configuration object) and then raise a `GenerationError` to signal that the call to `generate()` has failed. I've also added a test for this.

rupertford commented on src/psyclone/dynamo0p1.py:
```diff
@@ -1,10 +1,38 @@
-# whose members are identified at https://puma.nerc.ac.uk/trac/GungHo/wiki
-#-------------------------------------------------------------------------------
-# Author R. Ford STFC Daresbury Lab
+# -----------------------------------------------------------------------------
```
Above and beyond!

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5206,15 +5210,12 @@ def gen_code(self, parent):
-        if not self.module_inline:
-            parent.add(UseGen(parent, name=self._module_name,
-                              only=True, funcnames=[self._name]))
+        super(DynKern, self).gen_code(parent, create_arg_list.arglist)
```
Generation now done in base class. Better.

rupertford commented on src/psyclone/generator.py:
```diff
@@ -187,6 +194,10 @@ def generate(filename, api="", kernel_path="", script_name=None,
 
+    # Store Kernel-output options in our Configuration object
+    Config.get().kernel_output_dir = kern_out_path
+    Config.get().kernel_naming = kern_naming
```
This could theoretically raise an exception even though you limit the valid values. Should you check?

arporter replied:
I should and now I do :-)

rupertford commented on src/psyclone/ghproto.py:
```diff
@@ -1,17 +1,52 @@
-# whose members are identified at https://puma.nerc.ac.uk/trac/GungHo/wiki
-#-------------------------------------------------------------------------------
-# Author R. Ford STFC Daresbury Lab
+# -----------------------------------------------------------------------------
```
Above and beyond

rupertford commented on src/psyclone/gocean0p1.py:
```diff
@@ -1,11 +1,38 @@
-# ----------------------------------------------------------------------------
-# Author R. Ford STFC Daresbury Lab
-# Funded by the GOcean project
+# -----------------------------------------------------------------------------
```
Above and beyond.

rupertford commented on src/psyclone/gocean1p0.py:
```diff
@@ -919,11 +925,9 @@ def gen_code(self, parent):
-        if not self.module_inline:
-            parent.add(UseGen(parent, name=self._module_name, only=True,
-                              funcnames=[self._name]))
+        # Now sub-class specific setup is done, call the gen_code() method
```
Use of base class. Good.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -237,39 +237,40 @@ def __init__(self, api="", distributed_memory=None):
             raise GenerationError("PSyFactory: Internal Error: Unsupported "
                                   "api type '{0}' found. Should not be "
                                   "possible.".format(self._type))
+        return PSyClass(invoke_info)
```
Better thanks.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2909,11 +2927,30 @@ def coloured_text(self):
-                          funcnames=[self._name]))
+
+        # If kernel has been transformed then we rename it and write it to file
+        self.update()
```
Could this have a name that better explains what it does e.g. rename_and_write(), or just write_to_file()?

arporter replied:
It could and now it has.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2909,11 +2927,30 @@ def coloured_text(self):
-                          funcnames=[self._name]))
+
+        # If kernel has been transformed then we rename it and write it to file
+        self.update()
```
Should we avoid writing to file if we are (module) inlining?

arporter replied:
Oh yes, good point. I was initially worried that this would mean that we wouldn't be able to check whether or not a subsequent transformation of another instance of the same kernel gave the same result (which we forbid if the kernel-renaming scheme is "single"). However, I then realised that we only worry about that because we need to ensure that the correct versions of the kernels are called. If a kernel is module in-lined then it's scope is local and we don't have to worry about it clashing with other instances.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2988,6 +3025,198 @@ def ast(self):
+        name_idx = -1
+        fdesc = None
+        while not fdesc:
+            name_idx += 1
```
Presumably it does not make sense to use the namespace manager here?

arporter replied:
No, I don't think it does since we have to use the filesystem to manage our names.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2988,6 +3025,198 @@ def ast(self):
+                # this is part of a parallel build)
+                fdesc = os.open(
+                    os.path.join(Config.get().kernel_output_dir, new_name),
+                    os.O_CREAT | os.O_WRONLY | os.O_EXCL)
```
Nice.

arporter replied:
Thanks. The power of Stackexchange :-)

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2988,6 +3025,198 @@ def ast(self):
+        fll = FortLineLength()
+        new_kern_code = fll.process(str(self.ast))
+
+        if not fdesc:
```
Does it make sense (and is it possible) to check that with "unique" as the option we don't output a new file name if the content is the same?

arporter replied:
If you mean from within the test suite then I'm not sure it's possible without diving in to checking file-creation times. We test that an error is raised if the kernel does already exist and is different. Testing that we don't create something identical to something that we already have is difficult and won't alter correctness (since either way, we end up with the thing that we want).

rupertford replied:
OK.

rupertford commented on src/psyclone/psyGen.py:
```diff
@@ -2988,6 +3025,198 @@ def ast(self):
+        # contains the kernel subroutine as a type-bound procedure
+        orig_type_name = ""
+        new_type_name = ""
+        dtypes = walk_ast(self.ast.content, [Fortran2003.Derived_Type_Def])
```
Ooh, fparser2, how exciting.

arporter replied:
It's the future...

arporter commented on src/psyclone/configuration.py:
```diff
@@ -46,6 +46,13 @@
+# single = If any given kernel is transformed more than once then it is
+#          assumed that the same transformation is always applied and only
+#          one version of the transformed kernel is created.
+VALID_KERNEL_NAMING_SCHEMES = ["unique", "single"]
```
Oh yes, that's nicer. I should have asked you first time around as I wasn't entirely happy but couldn't think of anything better!

arporter commented:
I have now explicitly forbidden the transformation of in-lined kernels or the in-lining of tranformed kernels. This is documented and tested and is already the subject of #229.

arporter commented:
I finally have all modified code covered. I had to add more tests since some of the code I moved was not covered originally. I'm afraid the changes are fairly substantial so you probably need to review almost from scratch. Ready for second review from @rupertford.

rupertford commented on src/psyclone/ghproto.py:
```diff
@@ -1,17 +1,52 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
```
Please add yourself and then update to 2019!

arporter replied:
Done.

rupertford commented on src/psyclone/gocean0p1.py:
```diff
@@ -1,11 +1,38 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
```
Please add your authorship and 2019.

rupertford commented on src/psyclone/gocean0p1.py:
```diff
@@ -1,11 +1,38 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
```
Include your name and add 2019

arporter replied:
Done.

rupertford commented on src/psyclone/ghproto.py:
```diff
@@ -1,17 +1,52 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
```
Add your name and 2019

rupertford commented on src/psyclone/dynamo0p1.py:
```diff
@@ -1,10 +1,38 @@
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford, STFC Daresbury Lab
```
Add your name and 2019.

arporter replied:
Done.

arporter commented on src/psyclone/configuration.py:
```diff
@@ -265,6 +273,13 @@ def load(self, config_file=None):
+        # The scheme to use when re-naming transformed kernels.
+        # By default we ensure that each transformed kernel is given a
+        # unique name (within the specified kernel-output directory).
+        self._kernel_naming = "multiple"
```
Good point. Done.

arporter commented on src/psyclone/configuration.py:
```diff
@@ -459,14 +474,58 @@ def filename(self):
+        :rtype: str
+        '''
+        if not self._kernel_output_dir:
+            # We use the CWD if no directory has been specified
```
Yes, that's nicer.

arporter replied:
Well, it was nicer but it broke lots of tests as it meant that the CWD was only being set the first time the configuration module was loaded into the interpreter. I've therefore just *documented*
the fact that it is set to CWD.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -6104,8 +6100,13 @@ def indirection_dofmap(self, function_space, operator=None):
+        '''
+        :return: the kernel argument list. The generate function must be \
+                 called first.
+        :rtype: List of str
```
Oh yes.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -1290,8 +1290,12 @@ class DynamoPSy(PSy):
     invokes object (which controls all the required invocation calls).
     It also overrides the PSy gen method so that we generate dynamo
-    specific PSy module code. '''
+    specific PSy module code.
```
Done.

arporter commented on src/psyclone/generator.py:
```diff
@@ -296,6 +319,22 @@ def main(args):
+                  format(args.okern))
+            exit(1)
+        if not os.access(args.okern, os.W_OK):
+            print("Cannot write to specified kernel output directory ({0}).".
```
Done.

arporter commented:
I've addressed all comments but have noticed that running the test suite results in some Fortran files being created in the CWD. I shall investigate...

arporter commented:
Possibly it was linked to test failures. I can run the whole test suite (in parallel) and no Fortran files are generated. Ready for next review from @rupertford.

rupertford commented on src/psyclone/tests/kernel_transformation_test.py:
```diff
@@ -101,3 +109,294 @@ def test_accroutine():
+    dtypes = walk_ast(prog.content, [Fortran2003.Derived_Type_Def])
+    names = walk_ast(dtypes[0].content, [Fortran2003.Type_Name])
+    assert str(names[0]) == "continuity{0}_type".format(tag)
+    # TODO check compilation of code (needs Joerg's extension of compilation
```
Please add issue or pr number for this todo.

arporter replied:
Done.

arporter commented on src/psyclone/configuration.py:
```diff
@@ -87,6 +95,17 @@ class Config(object):
+    # importing of this module and the actual generation of code (e.g. as
+    # happens in the test suite), we do not store it here. Instead it
+    # is set in the Config.kernel_output_dir getter.
+    _default_kernel_naming = u"multiple"
```
Removed.

arporter commented on src/psyclone/configuration.py:
```diff
@@ -459,14 +489,57 @@ def filename(self):
+    @property
+    def kernel_naming(self):
+        '''
+        :returns: what naming scheme to use when writing transformed kernels
```
Done.

arporter commented on src/psyclone/generator.py:
```diff
@@ -237,6 +254,8 @@ def main(args):
     parser.add_argument(
         '-opsy', help='filename of generated PSy code')
+    parser.add_argument('-okern',
+                        help='directory in which to put transformed kernels')
```
Good idea - done.

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -32,7 +32,7 @@
 # POSSIBILITY OF SUCH DAMAGE.
 # -----------------------------------------------------------------------------
-# Authors R. Ford and A. R. Porter, STFC Daresbury Lab
+# Authors: R. Ford and A. R. Porter, STFC Daresbury Lab
```
Done.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3060,10 +3102,25 @@ def coloured_text(self):
-        parent.add(UseGen(parent, name=self._module_name, only=True,
-                          funcnames=[self._name]))
+
+        # If kernel has been transformed then we rename it. If it is *not*
```
Ooh, picky! :-)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3138,6 +3195,209 @@ def ast(self):
+        Writes the (transformed) AST of this kernel to file and resets the
+        'modified' flag to False. By default, the kernel is re-named
+        so as to be unique within the kernel output directory stored
+        within the configuration object. If config.kernel_naming is
```
I've elaborated.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3169,21 +3429,43 @@ def local_vars(self):
+        Abstract method to construct the class-specific argument list for a
+        kernel call. Must be overridden in API-specific sub-class.
+
+        :param parent: the parent (in the PSyIR) of the kernel call with \
```
I thought we'd agreed on PSyIR? I think adding an 'e' confuses the issue. Am happy to discuss :-)

rupertford replied:
For the record, we've agreed to use PSyIR from now on

rupertford commented:
@arporter. You don't seem to have pushed your changes.

arporter commented:
Oops. Changes are now pushed. I've also merged with master and fixed a bug that meant that repeated calls to psy.gen did not create the same code.

rupertford commented on doc/developers.rst:
```diff
@@ -1166,6 +1166,39 @@ exchange before the loop) or add existing halo exchanges after a loop
+Kernel Transformations
+++++++++++++++++++++++
+
+When a kernel is transformed we need to ensure that the PSy layer is
```
The first sentence is a bit of a throwaway i.e. it tells us something to worry about but does not say how this is managed. Either elaborate (how is this done and where is this done) or remove.

arporter replied:
Good spot. Removed.

rupertford commented:
Nearly there. Just a few issues and then it can go on master. Back to @arporter.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1,6 +1,7 @@
 # BSD 3-Clause License
 #
-# Copyright (c) 2017-18, Science and Technology Facilities Council
+# Copyright (c) 2017-189 Science and Technology Facilities Council
```
Good spot. Done.

arporter commented on doc/index.rst:
```diff
@@ -20,6 +20,7 @@
    nemo
    stub_gen
    line_length
+   fortran_naming_conventions
```
Done. I've also added a pointer from there to the PSyclone Script section of the docs.

arporter commented on doc/psyclone_script.rst:
```diff
@@ -255,3 +269,31 @@ that calls the PSyclone-generated code is responsible for initialising
 details on the use of this profiling functionality please see the
 :ref:`profiling` section.
+
+Naming of Transformed Kernels
```
Yes, that's better thanks.

arporter commented on doc/transformations.rst:
```diff
@@ -123,12 +126,52 @@ Kernels
+
+Since PSyclone is run separately for each Algorithm in a given
+application, ensuring that there are no name clashes for kernels in
+the application as a whole requires that some state is maintained
```
Done.

arporter commented:
From me to you @rupertford
